src/devices/cpu/tlcs900/tmp94c241.cpp

```diff
@@ -0,0 +1,1520 @@
+		WRMEML( m_xssp.d, m_pc.d );
+		m_xssp.d -= 2;
+		WRMEMW( m_xssp.d, m_sr.w.l );
+
+		/* Mask off any lower priority interrupts  */
+		m_sr.b.h = ( m_sr.b.h & 0x8f ) | ( level << 4 );
+
+		m_pc.d = RDMEML( 0xffff00 + vector );
+
+		m_cycles += 18;
+		m_prefetch_clear = true;
+
+		m_halted = 0;
+
+		/* Clear taken IRQ */
+		m_int_reg[ tmp94c241_irq_vector_map[irq].reg ] &= ~ tmp94c241_irq_vector_map[irq].iff;
+	}
+}
+
+
+//-------------------------------------------------
+//  tlcs900_handle_ad -
+//-------------------------------------------------
+
+void tmp94c241_device::tlcs900_handle_ad()
+{
+}
+
+
+//-------------------------------------------------
+//  tlcs900_handle_timers -
+//-------------------------------------------------
+
+void tmp94c241_device::tlcs900_handle_timers()
+{
+	uint32_t  old_pre = m_timer_pre;
+
+	if ( BIT(m_t16run, 7) ) /* prescaler is active */
+		m_timer_pre += m_cycles;
+
+	if ( BIT(m_t8run, 0) ) /* Timer 0 is running */
+	{
+		switch( m_t01mod & 3 ) /* T0_INPUT_CLOCK */
+		{
+		case 0:  /* TIO */
+			break;
+		case 1:  /* T1 */
+			m_timer_change[0] += ( m_timer_pre >> 3 ) - ( old_pre >> 3 );
+			break;
+		case 2:  /* T4 */
+			m_timer_change[0] += ( m_timer_pre >> 5 ) - ( old_pre >> 5 );
+			break;
+		case 3:  /* T16 */
+			m_timer_change[0] += ( m_timer_pre >> 7 ) - ( old_pre >> 7 );
+			break;
+		}
+
+		for( ; m_timer_change[0] > 0; m_timer_change[0]-- )
+		{
+			m_timer_8[0]++; /* UPCOUNTER_0 */
+			if ( m_timer_8[0] == m_treg_8[TREG0] )
+			{
+				if ( ((m_t01mod >> 6) & 3) == 0 ) /* TO1_OPERATING_MODE == MODE_8BIT_TIMER */
+					m_timer_change[1]++;
+
+				if ( (m_tffcr & 3) == 2 ) /* "FF1 Invert Enable" && "Invert by 8-bit timer 0" */
+					change_timer_flipflop( 1, FF_INVERT );
+
+				/* In 16bit timer mode the timer should not be reset */
+				if ( ((m_t01mod >> 6) & 3) != 1 ) /* TO1_OPERATING_MODE != MODE_16BIT_TIMER */
+				{
+					m_timer_8[0] = 0;
+					m_int_reg[INTET01] |= 0x08;
+					m_check_irqs = 1;
+				}
+			}
+		}
+	}
+
+	if ( BIT(m_t8run, 1) ) /* Timer 1 is running */
+	{
+		switch( (m_t01mod >> 2) & 3 ) /* T1_INPUT_CLOCK */
+		{
+		case 0:  /* TO0TRG */
+			break;
+		case 1:  /* T1 */
+			m_timer_change[1] += ( m_timer_pre >> 3 ) - ( old_pre >> 3 );
+			break;
+		case 2:  /* T16 */
+			m_timer_change[1] += ( m_timer_pre >> 7 ) - ( old_pre >> 7 );
+			break;
+		case 3:  /* T256 */
+			m_timer_change[1] += ( m_timer_pre >> 11 ) - ( old_pre >> 11 );
+			break;
+		}
+
+		for( ; m_timer_change[1] > 0; m_timer_change[1]-- )
+		{
+			m_timer_8[1] += 1; /* UPCOUNTER_1 */
+			if ( m_timer_8[1] == m_treg_8[TREG1] )
+			{
+				m_timer_8[1] = 0;
+				m_int_reg[INTET01] |= 0x80;
+				m_check_irqs = 1;
+
+				if ( (m_tffcr & 3) == 3 ) /* "FF1 Invert Enable" && "Invert by timer 1" */
+					change_timer_flipflop( 1, FF_INVERT );
+
+				/* In 16bit timer mode also reset timer 0 */
+				if ( ((m_t01mod >> 6) & 3) == 1 ) /* TO1_OPERATING_MODE == MODE_16BIT_TIMER */
+					m_timer_8[1] = 0;
+			}
+		}
```

Comment:
This seems to be repeated quite a bit.  Can you encapsulate at least some of it with a local lambda or something?

Reply:
OK. The remainder of the implementation of timers was generalized (at https://github.com/mamedev/mame/pull/13220/commits/9fcd79c530983d7893120408fa2e5c27257e53bf) with a couple local lambdas: One for the 8bit timers that can also optionally operate in pairs as 16 bit timers. And the other lambda is for the 16-bit only timers.